File: 백준/Silver/1802. 종이 접기/종이 접기.swift
let T = Int(readLine()!)!
for _ in 0..<T {
    solution()
}
func solution() {
    let paper = readLine()!.compactMap { $0.wholeNumberValue }
    if paper.count < 3 {
        print("YES")
    } else {
        let answer = divide(paper) ? "YES" : "NO"
        print(answer)
    }
    
}

func divide(_ arr: [Int]) -> Bool {
    if arr.count == 1 {
        return true
    }
    
    let mid = arr.count / 2
    let left = Array(arr[0..<mid])
    let right = arr[mid+1..<arr.count].map { 1-$0 }
    let revRight = Array(right.reversed())
    
    if left != revRight {
        return false
    } 
    
    return divide(left)
}

File: 백준/Silver/1802. 종이 접기/종이 접기.py

import sys
input = sys.stdin.readline

def reverse(arr):
    return [1-i for i in map(int, arr[::-1])]

def divide(arr):
    if len(arr) == 1:
        return True
    length = len(arr)//2
    
    left = arr[:length]
    right = arr[length+1:]
    revRight = reverse(right)
    
    if left != revRight:
        return False
    return divide(left)
    
def solution():
    paper = [i for i in map(int, input().rstrip())]
    if len(paper) < 3:
        print("YES")
    else:
        print("YES") if divide(paper) else print("NO")
    
T = int(input())
for i in range(T):
    solution()
